apollo-router/src/apollo_telemetry.rs

```diff
@@ -0,0 +1,553 @@
+static GRAPHQL_UNKNOWN_OPERATION_NAME: &str = "## GraphQLUnknownOperationName\n";
+
+fn normalize(op: Option<&opentelemetry::Value>, query: &str) -> String {
+    // If we don't have an operation name, we can't do anything useful
+    // with this query. Just return the appropriate error.
+    let op_name: String = match op {
+        Some(v) => v.as_str().into_owned(),
+        None => {
+            tracing::warn!("Could not identify operation name: {}", query);
+            return GRAPHQL_UNKNOWN_OPERATION_NAME.to_string();
+        }
+    };
+
+    let parser = Parser::new(query);
+    // compress *before* parsing to modify whitespaces/comments
+    let ast = parser.compress().parse();
+    tracing::debug!("ast:\n {:?}", ast);
+    // If we can't parse the query, we definitely can't normalize it, so
+    // just return the appropriate error.
+    if ast.errors().len() > 0 {
+        tracing::warn!("Could not parse query: {}", query);
+        return GRAPHQL_PARSE_FAILURE.to_string();
+    }
+    let doc = ast.document();
+    tracing::debug!("{}", doc.format());
+    tracing::trace!("looking for operation: {}", op_name);
+    let mut required_definitions: Vec<_> = doc
+        .definitions()
+        .into_iter()
+        .filter(|x| {
+            if let ast::Definition::OperationDefinition(op_def) = x {
+                return match op_def.name() {
+                    Some(v) => v.text() == op_name,
+                    None => op_name == "-",
+                };
+            }
+            false
+        })
+        .collect();
+    tracing::debug!("required definitions: {:?}", required_definitions);
+    if required_definitions.len() != 1 {
+        tracing::warn!("Could not find required single definition: {}", query);
+        return GRAPHQL_VALIDATION_FAILURE.to_string();
+    }
+    let required_definition = required_definitions.pop().unwrap();
+    tracing::debug!("required_definition: {:?}", required_definition);
+    // XXX Somehow find fragments...
+    let def = required_definition.format();
+    format!("# {}\n{}", op_name, def)
+}
+
+struct DurationHistogram {
+    buckets: Vec<i64>,
+    entries: u64,
+}
+
+static EXPONENT_LOG: OnceCell<f64> = OnceCell::new();
+
+impl DurationHistogram {
+    const DEFAULT_SIZE: usize = 74; // Taken from TS implementation
+    const MAXIMUM_SIZE: usize = 383; // Taken from TS implementation
+
+    fn new(init_size: Option<usize>) -> Self {
+        Self {
+            buckets: vec![0; init_size.unwrap_or(DurationHistogram::DEFAULT_SIZE)],
+            entries: 0,
+        }
+    }
+
+    fn duration_to_bucket(duration: Duration) -> usize {
+        // If you use as_micros() here to avoid the divide, tests will fail
+        // Because, internally, as_micros() is losing remainders
+        let log_duration = f64::log2(duration.as_nanos() as f64 / 1000.0);
+        let unbounded_bucket =
+            f64::ceil(log_duration / EXPONENT_LOG.get_or_init(|| f64::log2(1.1)));
```

Comment:
why is EXPONENT_LOG in a static cell? I don't see it modified anywhere else, it could be a constant, right?

Reply:
Good catch. I made it a constant. done: d5b2d7f71b63660e1a50a6698003bec94e3f90b7

Reply:
Ah, now I remember (or at least the compiler remembers):

```
warning: a `const` item should never be interior mutable
   --> apollo-router/src/apollo_telemetry.rs:390:5
    |
390 |     const EXPONENT_LOG: OnceCell<f64> = OnceCell::new();
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(clippy::declare_interior_mutable_const)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#declare_interior_mutable_const

warning: a `const` item with interior mutability should not be borrowed
   --> apollo-router/src/apollo_telemetry.rs:404:28
    |
404 |             log_duration / DurationHistogram::EXPONENT_LOG.get_or_init(|| f64::log2(1.1)),
    |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Reply:
my point was less about static/const, but more about the cell. Why does it need interior mutability, since it's assigned only once, with a constant, and never modified afterwards?

Reply:
confusion resolved: log2 isn't a const fn(), so we can't have const value derived from it. Performant option: manually calculate log2(1.1) (ANSWER: 0.13750352375) and include this with a comment as a const value in the code. We can update to the calculated value when/if f64::log2() becomes const.
done: 2b01014ee6048ace2fd7cce3b679cc87fc60c752